src/main/java/emu/grasscutter/game/ability/AbilityManager.java

```diff
@@ -152,22 +160,65 @@ private void handleGenerateElemBall(AbilityInvokeEntry invoke) throws InvalidPro
 		this.player.getEnergyManager().handleGenerateElemBall(invoke);
 	}
 	
-	private void invokeAction(AbilityModifierAction action, GameEntity target, GameEntity sourceEntity) {
+    private class HealData {
+        public String avatar = "";
+        public int type= 0;
+        public float factor = 0;
+        public float base = 0;
+
+        public HealData(String _avatar, int _type, float _factor, float _base) {
+            avatar = _avatar;
+            type = _type;
+            factor = _factor;
+            base = _base;
+        }
+    }
+
+	private void invokeAction(AbilityModifierAction action, GameEntity target, GameEntity sourceEntity, String modifierString) {
 		switch (action.type) {
 			case HealHP -> {
-				if (action.amount == null) {
-					return;
-				}
-				
+                ArrayList<HealData> healDataList = new ArrayList();
```

Comment:
While this is a step in the right direction, variables from HealData are dynamic and recieved from the client so there is no need to hard code every character's healing amount.